### PR TITLE
Add LA scraper

### LINKED ISSUE #79: Add LA scrape
http://www.laworks.net/Downloads/Downloads_WFD.asp

Links to PDFs for each year (see the Rapid Response section of the page).

### PR DESCRIPTION
**Pull request by chriszs**
Getting on the board in the home of the Big Easy while [Chi-Town rests](https://github.com/biglocalnews/warn-scraper/pull/358).

## Steps to test
```sh
python -m warn.cli LA -l DEBUG
```

Closes #79

### CODE REVIEW

**Comment by chriszs** [comment]
Currently places address information in the company name field as do the PDFs. Only way to reliably distinguish seems to be to rely on difference in text weight:

Screen Shot 2022-01-22 at 1 44 46 PM

Looks like I may be able to do that by querying the page for characters within the bounding boxes of `TableFinder`.

File: warn/scrapers/la.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                    output_row = [_clean_text(cell) for cell in row]
+
+                    # Skip headers on all but first page of first PDF
+                    if (pdf_index == 0 and page_index == 0) or row_index != 0:
+                        output_rows.append(output_row)
+
+    # Patch in missing header value
+    output_rows[0][3] = "Employees Affected"
+
+    data_path = Path(data_dir, f"{state_code}.csv")
```
Comment (palewire) [inline warn/scrapers/la.py:73]:
```suggestion
    data_path = data_dir, / f"{state_code}.csv"
```
A little trimski

**Comment by chriszs** [comment]
That method works for all but 2020 and 2021:
Screen Shot 2022-01-22 at 3 05 57 PM

So, need a fallback plan. Most contain a number somewhere on the first address line, usually at the beginning. A naive approach splits on that and then special cases the small handful of exceptions.

**Comment by chriszs** [comment]
I wonder if I can use words in the bolded examples to learn how to classify the non-bolded examples.

File: warn/scrapers/la.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    html = page.text
+
+    cache.write(cache_key, html)
+
+    document = BeautifulSoup(html, "html.parser")
+    links = document.find_all("a")
+
+    pdf_urls = [
+        f"{base_url}{link['href']}" for link in links if "WARN Notices" in link.text
+    ]
```
Comment (palewire) [inline warn/scrapers/la.py:49]:
```suggestion
    # Fire up the cache
    cache = Cache(cache_dir)

    # Download the root page
    base_url = "https://www.laworks.net/"
    url = f"{base_url}Downloads/Downloads_WFD.asp"
    page = utils.get_url(url)
    html = page.text

    # Save it to the cache
    state_code = "la"
    cache_key = f"{state_code}/Downloads_WFD.html"
    cache.write(cache_key, html)

    # Parse out the PDF links
    document = BeautifulSoup(html, "html.parser")
    links = document.find_all("a")
    pdf_urls = [
        f"{base_url}{link['href']}" for link in links if "WARN Notices" in link.text
    ]
```
A little tidying

File: warn/scrapers/la.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        with pdfplumber.open(pdf_path) as pdf:
+            for page_index, page in enumerate(pdf.pages):
+                rows = page.extract_table()
+
+                for row_index, row in enumerate(rows):
+                    output_row = [_clean_text(cell) for cell in row]
+
+                    # Skip headers on all but first page of first PDF
+                    if (pdf_index == 0 and page_index == 0) or row_index != 0:
+                        output_rows.append(output_row)
```
Comment (palewire) [inline warn/scrapers/la.py:68]:
```suggestion
                # Loop through the rows
                for row_index, row in enumerate(rows):
                    # Skip headers on all but first page of first PDF
                    if pdf_index > 0 and row_index == 0:
                        print(row)
                        logger.debug(f"Skipping header row on PDF {pdf_index+1} page {page_index+1}")
                        continue

                    # Extract data
                    output_row = [_clean_text(cell) for cell in row]

                    # Write row
                    output_rows.append(output_row)
```
This doesn't solve your "sometimes there a header, and sometimes there isn't a header` problem, but it might set up a clearer conditional clause where you can experiment with the fix.

File: warn/scrapers/la.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    # Patch in missing header value
+    output_rows[0][3] = "Employees Affected"
+
+    data_path = data_dir / f"{state_code}.csv"
+    utils.write_rows_to_csv(output_rows, data_path)
+
+    return data_path
```
Comment (palewire) [inline warn/scrapers/la.py:76]:
```suggestion
   # Write out to CSV
    data_path = data_dir / f"{state_code}.csv"
    utils.write_rows_to_csv(output_rows, data_path)

    # Return the path
    return data_path
```
A little more docstringin'

File: warn/scrapers/la.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    Keyword arguments:
+    text -- the text to clean
+
+    Returns: the cleaned text
+    """
+    if text is None:
+        return ""
+    # Standardize whitespace
+    return re.sub(r"\s+", " ", text)
```
Comment (palewire) [inline warn/scrapers/la.py:91]:
```suggestion
    # Replace None with an empty string
    if text is None:
        return ""

    # Standardize whitespace
    return re.sub(r"\s+", " ", text)
```

**Review by palewire — CHANGES_REQUESTED** [review]

File: warn/scrapers/la.py (hunk 4 above)
Comment (chriszs) [reply warn/scrapers/la.py:76]: I can go along with the style for consistency, but when the comment says exactly what the line below it already says, I've been converted to the idea that's [a mistake](https://www.red-gate.com/simple-talk/opinion/opinion-pieces/fighting-evil-code-comments-comments/) and that in general comments are a signal to better factor the code.

**Comment by chriszs** [comment]
Okay now uses the bolding to separate out the company name and address when it's available and a pattern-based heuristic when there is no bold text. Also separates out notes like "UPDATE" and "WARN RESCINDED." Also deals with one row that ran over on to a second page. Doesn't guarantee perfect tidiness, but uses what it has to give you something closer.

Screen Shot 2022-01-23 at 9 44 53 PM

File: warn/scrapers/la.py [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+            # Process the PDF
+            rows = _process_pdf(pdf_path)
+            output_rows.extend(rows)
+
+    # Insert a header row with clean column names.
+    # We are here assuming that the columns don't change between years
+    # and that one that contains "Employees Affected" will be clean.
+    output_rows = [list(filter(_is_clean_header, output_rows))[0]] + list(
+        filter(_is_not_header, output_rows)
+    )
```
Comment (palewire) [inline warn/scrapers/la.py:68]: Picky thing. Can we name the first `output_rows` something else? Strictly speaking, I think it starts out as more than what we are going to output, and then gets filtered down at the tail end here. On my first read through the code it tripped me up a little.

File: warn/scrapers/la.py [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+    # Download the root page
+    url = f"{base_url}Downloads/{file_base}.asp"
+    html = utils.get_url(url).text
+
+    # Save it to the cache
+    cache_key = f"{state_code}/{file_base}.html"
+    cache.write(cache_key, html)
+
+    # Parse out the links to WARN notice PDFs
+    links = _parse_links(html)
```
Comment (palewire) [inline warn/scrapers/la.py:49]: If this function is just a couple lines of code and is only getting called once, is it necessary? Could we do it inline?

**Comment by chriszs** [comment]
One potential enhancement here would be to better deal with these multi-line cells, usually containing multiple updates for the same business:
Screen Shot 2022-01-24 at 12 23 53 PM
Screen Shot 2022-01-24 at 12 27 32 PM

File: warn/scrapers/la.py (hunk 7 above)
Comment (chriszs) [reply warn/scrapers/la.py:49]: We could. There are maybe two competing code clarity demands:
1. Implementation hiding (do you really need to know the details of how I'm parsing the links at a high level? ideally, we just call it by a name and use that)
2. Complexity reduction (do you really need to make me go seek out a function definition to find out how this works? ideally, it's all simple enough we can do it inline)

Add to that the possibility of refactoring the definition of how we parse links without having to change the top level flow.

I'm opting for more of 1 in this case, because this got more complicated than most other scrapers, and I wanted to keep the top level flow simple, but I did course correct during this refactor and get rid of a couple one-line functions that were doing something simple, because they obscured more than they enlightened. Could do the same here if you think that's wisest.

Comment (palewire) [reply warn/scrapers/la.py:49]: If you feel strongly about the approach, I'm fine with keeping it. My lizard brain was living more in the DRY world where you only need a function if you are going to reuse it.

File: warn/scrapers/la.py (hunk 6 above)
Comment (chriszs) [reply warn/scrapers/la.py]: Sure, done.

File: warn/scrapers/la.py [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+                        and len(output_rows) > 0
+                    ):
+                        for column_index, cell in enumerate(row):
+                            if column_index < len(output_rows[len(output_rows) - 1]):
+                                output_rows[len(output_rows) - 1][column_index].extend(
+                                    cell
+                                )
+                    # Otherwise, append the row
+                    else:
+                        output_rows.append(row)
```
Comment (chriszs) [inline warn/scrapers/la.py:110]: Feels like this could be better factored.

Comment (palewire) [reply warn/scrapers/la.py:110]: As someone who doesn't understand the core issue, it's def. hard to parse. I don't have the perfect solution off the top of my head, but I think so that techniques like breaking out clearly named true/false functions, and breaking it up into a series of if statements rather than one mega clause, tends to help.

Comment (chriszs) [reply warn/scrapers/la.py:110]: It's handling this one line:
Screen Shot 2022-01-28 at 1 07 30 AM

Comment (chriszs) [reply warn/scrapers/la.py:110]: But if you don't understand from looking at the code, I haven't done my job yet.

Comment (palewire) [reply warn/scrapers/la.py:110]: PDFs are the devil!

Comment (chriszs) [reply warn/scrapers/la.py]: Tried to clean this up.

File: warn/scrapers/la.py (hunk 7 above)
Comment (chriszs) [reply warn/scrapers/la.py]: I killed this function.

**Comment by palewire** [comment]
Thanks for your help here, @chriszs. You're in!
